server/Database.hs

```diff
@@ -0,0 +1,187 @@
+import Crypto.Fido2.Protocol
+  ( CredentialId (CredentialId),
+    URLEncodedBase64 (URLEncodedBase64),
+    UserId (UserId),
+  )
+import qualified Crypto.Fido2.Protocol as Fido2
+import qualified Data.Binary as Binary
+import qualified Database.SQLite.Simple as Sqlite
+
+type Connection = Sqlite.Connection
+
+newtype Transaction = Transaction Sqlite.Connection
+
+connect :: IO Sqlite.Connection
+connect = do
+  conn <- Sqlite.open "users.sqlite3"
+  Sqlite.execute conn "pragma foreign_keys = on;" ()
+  pure conn
+
+initialize :: Sqlite.Connection -> IO ()
+initialize conn = do
+  Sqlite.execute
+    conn
+    " create table if not exists users                                         \
+    \ ( id           blob primary key                                          \
+    \ , username     text not null unique                                      \
+    \ , display_name text not null                                             \
+    \ , created      text not null                                             \
+    \                default (strftime('%Y-%m-%dT%H:%M:%fZ', 'now'))           \
+    \ );                                                                       "
+    ()
+  Sqlite.execute
+    conn
+    " create table if not exists attested_credential_data                      \
+    \ ( id               blob    primary key                                   \
+    \ , user_id          blob    not null                                      \
+    \ , public_key_x     blob    not null                                      \
+    \ , public_key_y     blob    not null                                      \
+    \ , created          text    not null                                      \
+    \                    default (strftime('%Y-%m-%dT%H:%M:%fZ', 'now'))       \
+    \ , foreign key (user_id) references users (id)                            \
+    \ );                                                                       "
+    ()
+  Sqlite.execute
+    conn
+    " create index if not exists                                               \
+    \ ix_attested_credential_data_user_id                                      \
+    \ on attested_credential_data(user_id);                                    "
+    ()
+
+-- | Run an action using `Sqlite.withTransaction`.
+--
+-- If exceptions occur within the transaciton, the transaction is aborted
+-- with `ROLLBACK TRANSACTION`. Otherwise, the transaction is committed using
+-- `COMMIT TRANSACTION`.
+--
+-- This ensures that we always close our transactions and don't leave them
+-- open when exceptions occur.
+withTransaction :: Sqlite.Connection -> (Transaction -> IO a) -> IO a
+withTransaction conn action = Sqlite.withTransaction conn (action (Transaction conn))
+
+begin :: Sqlite.Connection -> IO Transaction
+begin conn = do
+  Sqlite.execute conn "begin;" ()
+  pure $ Transaction conn
+
+commit :: Transaction -> IO ()
+commit (Transaction conn) = Sqlite.execute conn "commit;" ()
```

Comment:
Are these still needed now that there is `withTransaction`?

Reply:
Thanks! Removed these